src/lightning/fabric/utilities/throughput.py

```diff
@@ -0,0 +1,441 @@
+    +--------------------------+-----------------------------------------------------------------------------------+
+    | time                   | Total elapsed time                                                                  |
+    +--------------------------+-----------------------------------------------------------------------------------+
+
+    Example::
+
+        throughput = Throughput()
+        t0 = time()
+        for i in range(1000):
+            do_work()
+            if torch.cuda.is_available(): torch.cuda.synchronize()  # required or else time() won't be correct
+            throughput.update(time=time() - t0, samples=i)
+            if i % 10 == 0:
+                print(throughput.compute())
+
+    Notes:
+        - The implementation assumes that devices FLOPs are all the same as it normalizes by the world size and only
+          takes a single ``available_flops`` value.
+        - items_per_sec, flops_per_sec and MFU do not account for padding if present. We suggest using
+          samples_per_sec or batches_per_sec to measure throughput under this circumstance.
+
+    Args:
+        available_flops: Number of theoretical flops available for a single device.
+        world_size: Number of devices available across hosts. Global metrics are not included if the world size is 1.
+        window_size: Number of batches to use for a rolling average.
+        separator: Key separator to use when creating per-device and global metrics.
+
+    """
+
+    def __init__(
+        self, available_flops: Optional[float] = None, world_size: int = 1, window_size: int = 100, separator: str = "/"
+    ) -> None:
+        self.available_flops = available_flops
+        self.separator = separator
+        assert world_size > 0
+        self.world_size = world_size
+
+        # throughput is computed over a window of values. at least 2 is enforced since it looks at the difference
+        # between the first and last elements
+        assert window_size > 1
+        # custom class instead of `deque(maxlen=)` because it's easy for users to mess up their timer/counters and log
+        # values that do not increase monotonically. this class will raise an error if that happens.
+        self._samples: _MonotonicWindow[int] = _MonotonicWindow(maxlen=window_size)
+        self._time: _MonotonicWindow[float] = _MonotonicWindow(maxlen=window_size)
+        self._lengths: _MonotonicWindow[int] = _MonotonicWindow(maxlen=window_size)
+        self._flops: Deque[int] = deque(maxlen=window_size)
+
+    def update(
+        self, *, time: float, samples: int, lengths: Optional[int] = None, flops_per_batch: Optional[int] = None
+    ) -> Self:
+        """Update throughput metrics.
+
+        Args:
+            time: Total elapsed time in seconds. It should monotonically increase by the iteration time with each
+                call.
+            samples: Total samples seen per device. It should monotonically increase by the batch size with each
```

Comment:
```suggestion
            sample_counter: Total samples seen per device. It should monotonically increase by the batch size with each
```

Reply:
It's the number of samples. `_counter` would suggest this is some other entity that is counting.